### ISSUE TITLE
Stack frame names have `[Unoptimized] ` prefix

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#35288]
I'm not sure if this is deliberate, but it seems a bit strange. We use the `name` from the `source` in `StackFrame` to show in the IDE, and in a recent dev build they've started being prefixed with `[Unoptimized]`:

<img width="489" alt="screen shot 2018-11-29 at 4 15 21 pm" src="https://user-images.githubusercontent.com/1078012/49235537-30bb0c00-f3f2-11e8-841a-240fe4337529.png">

It seems like this would be better as a flag to let the editor choose how to display it rather than putting it here (or at least, having it on the end so it doesn't truncate the visible method names as much).

(@a-siva is this something you'd know about?)

**Comment by rmacnak-google** [comment]
Where are you reading this from? There is no type called 'StackFrame' exposed over the VM service.

**Comment by DanTup** [comment]
Sorry, I should've been more explicit. We read the `name` field from a `Frame`'s `code`., for ex. `getStack().frames[x].code.name`. I don't know if this is the correct thing for the IDE to show in the call stack list, but it seemed appropriate (up until now, at least).

If there's something better, I can easily change this though.

**Comment by a-siva** [comment]
getStack().frames[x].function.name should give you the function name

**Comment by DanTup** [comment]
The value in `.function.name` doesn't seem to be as descriptive as the one from `.code`, for example:

```
frame.code.name:     main.<anonymous closure>.<anonymous closure>
frame.function.name: <anonymous closure>
```

I think that IntelliJ also uses `.code.name` if I'm looking in the right place (@pq?):

https://github.com/JetBrains/intellij-plugins/blob/5fb99c284ae67a4895da6120a349b270c9149699/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/frame/DartVmServiceStackFrame.java#L79

So I'm not sure what to do - I think changing it to `.function` would be less useful (and the issue would probably still exist for IntelliJ). Do we need to have the prefix in `.code.name` or could we change it back (or add it as its own field or something)?

**Comment by rmacnak-google** [comment]
No, Code objects should not be displayed. Even before this change, Code.name would sometimes be prefixed with a star (which meant optimized). The appropriate name to display to users is Function.name, which can be qualified if desired by pretending the name of Function.owner.

**Comment by DanTup** [comment]
I tried prefixing with the .owner's names (recursively), but I found instances where `frame.func.name` and `frame.func.owner.name` are the same, so I end up with duplication (that wasn't there in the previous code):

<img width="351" alt="screen shot 2018-12-03 at 5 31 49 pm" src="https://user-images.githubusercontent.com/1078012/49390619-59f4d880-f721-11e8-96db-ea299b38c209.png">


I'm not really sure what is the correct representation for these since they're kind cryptic.

@pq @devoncarew maybe if you figure out what you're going to do for IntelliJ I can just borrow it? 😁

**Comment by mraleph** [comment]
Implicit closures are objects used to represent tear-offs, so they have the same name as the original non-torn off method. I think you can detect this case easily by checking the kind.

**Comment by DanTup** [comment]
Thanks! Is there an easy way to tell what the full list of kinds are (or whether any others may be the same as their parents like this)? It doesn't look like they're in [this doc](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#function) - or at least, I can't see `ImplicitClosureFunction` in there.
